Major tesseract improvements

erikarvstedt opened the pull request:
1. Decouple Tesseract from the tessdata language corpus, thus allowing a
   lightweight installation for users that want to dynamically provide their
   language data via environment vars.
2. Most importantly, this allows users to override the supported languages without
   triggering a time-consuming recompilation. This brings a significant speed boost.
3. Provide all languages as individual derivations.

Here's a quick demo:

```nix
{ pkgs ? (import <nixpkgs> {})}:

with pkgs;
{
  inherit tesseract tesseract4;

  # Assemble the tessdata from individual language derivations without
  # downloading the whole tessdata corpus.
  # This downloads in a just a few seconds. Previously, this triggered a full
  # tessdata download (1.2 GB unzipped) + a full recompilation.
  tesseractWithCustomLangs = tesseract.override {
    enableLanguages = [ "eng" "fra" ];
  };

  # Low-level custom tessdata
  tesseractWithUserTessdata = tesseract.override {
    tessdata = [
      tesseract.languages.eng
      (fetchurl {
        url = "https://github.com/tesseract-ocr/tessdata/blob/3cf1e2df1fe1d1da29295c9ef0983796c7958b7d/tel.traineddata";
        sha256 = "1h4xn6ccd24hv4ps5kg53vk1cxfcqk6w4v1k48lps8f3zpn5aix8";
      })
    ];
  };
}
```

## Obsolete, historical content (for the archives)

### To be discussed:
I'd recommend that we provide all tessdata languages as individual derivations.
Bonuses:
  - Increased user friendliness.
  - Languages are no longer downloaded and stored twice in /nix/store
    (once as an individual derivation, once as part of languages.all).
    Although for `--optimise`d stores, the storage concerns are irrelevant.
  - We can get rid of some special-case handling in the tessdata definition in `default.nix`.

The extra maintenance burden is negligible due to the `get-language-hashes.sh` helper.

**Edit 2:** 
We've arrived at the following:
Keep a simple, single fixed output drv combining all languages that's used by `pkgs.tesseract` and is cached by Hydra, as before. This keeps eval times low.
Provide individual derivations for all languages, but don't cache them with Hydra. These drvs are downloaded individually from the original repo source when `tesseract.enableLanguages` is modified.

**Edit:** 
I've added a [commit](https://github.com/NixOS/nixpkgs/pull/52379/commits/51af0271b0f19cca9824ef64207cf8233c846971) that implements separate lang drvs.
The evaluation time of `nixpkgs.tesseract` (which includes all languages) is increased by 6.8% (361ms vs 338ms, via [hyperfine](https://github.com/sharkdp/hyperfine)) but I think the added flexibility and simplicity is worth the little extra eval time.

Here's a quick script to verify that the new tessdata files are identical to the old ones:
```
new=$(nix-build --no-out-link -E '(import <nixpkgs-this-PR>  { config = {}; overlays = []; }).tesseract')
old=$(nix-build --no-out-link -E '(import <nixpkgs-unstable> { config = {}; overlays = []; }).tesseract')
for lang in $new/share/tessdata/*.traineddata; do
    oldLang=$old/share/tessdata/$(basename $lang)
    if ! cmp $lang $oldLang; then
        echo $lang differs
        exit 1
    fi
done
```

7c6f434c commented:
Won't your approach to getting the language hashes fail if the language has already been downloaded before (but changed upstream)?

erikarvstedt commented:
Good point! I've [switched to `nix-prefetch-url`](https://github.com/NixOS/nixpkgs/pull/52379/commits/f2d64f0341a19d895459ca1a43362b67957bbcee).

7c6f434c commented on pkgs/applications/graphics/tesseract/default.nix:
```diff
@@ -1,67 +1,91 @@
 # to make tessdata a fixed output derivation.
 , enableLanguagesHash ? (if enableLanguages == null # all languages
-                         then "1h48xfzabhn0ldbx5ib67cp9607pr0zpblsy8z6fs4knn0zznfnw"
+                         then "11bi1hj2ihqrgvi9cam8mi70p4spm3syljkpnbglf4s8jkpfn15a"
```
Is this default ever used? If `enableLanguages` is `null`, `languages.all` is passed through without any extra processing…

erikarvstedt replied:
Your're right, it's never used.
If we switch to single lang derivations only, we can get rid of `enableLanguagesHash ` completely. Otherwise, we should remove the unused default value.

7c6f434c replied:
I think if selected-languages case works, killing it might be suboptimal: there are some substantially  two-language texts.

erikarvstedt replied:
I meant that if we switch to single lang derivations only, `tessdata` could be just a list of fixed-output lang derivations, so no extra output hash would be needed.

7c6f434c commented on pkgs/applications/graphics/tesseract/default.nix:
```diff
@@ -1,67 +1,91 @@
+          buildCommand = ''
+            mkdir $out
+            cd ${languages.all}
+            cp ${stdenv.lib.concatMapStringsSep " " (x: x + ".traineddata") enableLanguages} $out
```
Is there anything large enough _not_ copied in the downloaded language packs? If no, you could just `ln -s`, I guess?

erikarvstedt replied:
Yes. If we just need a single exotic language, like `cym`, all other languages (1.2 GB) are needlessly kept in the store.
The old behaviour was also to copy the files.

7c6f434c replied:
Oops, my bad, I thought each language is copied from the individual language pack.

erikarvstedt replied:
In case you missed it: I've added a line to my previous reply.

7c6f434c replied:
I don't understand that `if` now. «If we mention all languages, use individual derivations instead of `all`»?

If just a list of derivations selected by language names would work without copying, that sounds better than that extra composition step.

erikarvstedt replied:
Currently we're not providing all tessdata languages as separate drvs, just a selection of the most popular ones. 
So for more obscure langs we have to tap into `languages.all`.
Here's a translation of the second `if`-statement:
```
If all languages are available as separate drvs:
    build a list of drvs
else:
    build a new tessdata drv by copying all requested languages from languages.all
```

erikarvstedt replied:
Note, again, that this is hugely simplified if switch to seperate drvs for all langs. :smiley_cat:

7c6f434c replied:
Now I understand what «switching to individual derivations for all languages» actually mean… Yes, sounds like a good idea. I thought it was about less support for multi-language builds.

erikarvstedt replied:
Alright, sorry if I wasn't perfectly clear about that.
I've added a [commit](https://github.com/NixOS/nixpkgs/pull/52379/commits/51af0271b0f19cca9824ef64207cf8233c846971) that implements separate lang drvs.
The evaluation time of `nixpkgs.tesseract` (which includes all languages) is increased by 6.8% (361ms vs 338ms, via [hyperfine](https://github.com/sharkdp/hyperfine)) but I think the added flexibility and simplicity is worth the little extra eval time. What do you think?

7c6f434c replied:
Maybe default to the old list of languages? Would it bring the time back down?

erikarvstedt replied:
With the current approach we're evaluating one `fetchurl` drv for each language (107 total) and pass their store paths to the builder.

If we bring back a single fixed output drv for all langs, then the performance will be the same as before.

7c6f434c replied:
Hm. Now I wonder if it is better to have a medium-size list of languages by default, or just download the huge all-languages tarball…

7c6f434c replied:
(Yes, this might show you that it is not you being dense, this is also me being unsure and therefore not always picking words well. I guess I will trust your choice in this question, given that building a local three-language version will be cheap)

erikarvstedt replied:
Ah, my doubts were confirmed: Caching the 'all languages' tessdata with a fixed output drv that links all lang drvs into a single folder is not possible. That's because nix always fully evaluates all derivation attributes, even for fixed output drvs. So all lang drvs that are passed to the builder will still be evaluated.

We have the following options now:
1. Accept the slightly worse perf.
2. Cache 'all languages' with a drv that fetches the whole tessdata repo, like before. But langs are stored twice this way. (Hydra doesn't seem to use `nix-store --optimise`).
3. As proposed by you, restrict default langs for `nixpkgs.tesseract` to a smaller subset. But this breaks backwards compability. Previously, the full tessdata has always been included .

erikarvstedt replied:
I'm actually tending towards 3. What do you think?

7c6f434c replied:
There is also the option 2.1, which fetches the whole repository and sets empty `meta.hydraPlatforms` for the individual languages. Then only English is stored twice on Hydra (used for VM tests). Arguably, we could even avoid keeping any language except English on Hydra… (option 2.2)

erikarvstedt replied:
Please see **"Edit 2"** in the main PR description for a short summary of our current approach. I think we've hit the sweet spot.

7c6f434c commented on pkgs/applications/graphics/tesseract/get-language-hashes.sh:
```diff
@@ -10,14 +14,29 @@
+fi
 
-nixSrc=$(sed "s/TESSDATA_REV/$tessdataRev/" <<'EOF'
+nixSrc=$(sed -e "s/TESSDATA_REV/$tessdataRev/" -e "s/LANGUAGE_CODES/$langCodesExpr/" <<'EOF'
```
Is it just me, or would just using shell (and exporting the list from Nix using `concatStringsSep, if needed`) be simpler in the current approach?

erikarvstedt replied:
Could you clarify a bit what you mean by this?

erikarvstedt replied:
Yep, that's infinitely more elegant, thanks for pointing it out! [Fixed.](https://github.com/NixOS/nixpkgs/pull/52379/commits/01e3738ddc40fe331649bad50bf2a27ebe3f82d5)

7c6f434c commented on pkgs/top-level/all-packages.nix:
```diff
@@ -19486,6 +19486,7 @@ in
   termtosvg = callPackage ../tools/misc/termtosvg { };
 
   tesseract = callPackage ../applications/graphics/tesseract { };
+  tesseractWithoutData = callPackage ../applications/graphics/tesseract { enableLanguages = false; };
```
1. Maybe just use `[]` for the no-languages case?

2. It is already available as both `tesseract.tesseractWithoutData` and (if you agree with point 1) `tesseract.override {enableLanguages = [];}`, does it need a top-level attribute or maybe just a top-level _comment_?

```
# you can pass `enableLanguages` to `tesseract` to install tessearct with fewer supported languages and save space,
# or even pass `enableLanguages = []` and then supply the language packs via environment.
```

erikarvstedt replied:
1. I have no preference, I'll let you decide.
2. We need a top-level attr so that the no-languages drv is cached. Without it, we would incur a full compilation when `enableLanguages` is customized.

7c6f434c replied:
I think `[]` reduces the number of types (list or null seems to be a reasonable union type, list of null or false boolean should provide some value). So I would slightly prefer `[]`

Given the size of the compiled part (less than a single language model), maybe symlink the files from there and only copy the files that need to be edited to replace the path? I think `sed` can even break symlinks automatically in the inplace mode. Then the binary build becomes a dependency of the toplevel attribute.

(I am not sure if Hydra actually keeps the build depepndencies — then the entire point is moot — but even if not, not a large problem)

erikarvstedt replied:
I was just typing up exactly what you proposed here. :smile_cat: 
With this option, the build script of `tesseractWithData` will be a bit more complex, but we lose the slightly unelegant extra top-level attr. I think that's preferable.
I'll have breakfast and then implement it. Thanks for your very thoughtful feedback!

erikarvstedt replied:
[Done.](https://github.com/NixOS/nixpkgs/pull/52379/commits/b02785bfc4b61b1f1e4b09fabba49d0cd1a77fba)

7c6f434c commented:
Just in case @viric has any comments, I mention him (he is less active these days, though).

7c6f434c commented on pkgs/applications/graphics/tesseract/default.nix:
```diff
@@ -1,67 +1,82 @@
+      description = "OCR engine";
+      homepage = https://github.com/tesseract-ocr/tesseract;
+      license = stdenv.lib.licenses.asl20;
+      maintainers = with stdenv.lib.maintainers; [viric];
```
You might want to add yourself as a maintainer.

erikarvstedt replied:
[Done.](https://github.com/NixOS/nixpkgs/pull/52379/commits/412a5b33213a9bc07498b43beca81a5ec01dc5e4)

erikarvstedt commented:
There's still one very serious problem to overcome.
Sorry for not mentioning it earlier, I've just come to realize its full scope:
`pkgs.tesseract.overrideAttrs` is not backwards-compatible, because the overridden attrs are not passed on to the underlying no-languages tesseract drv.

Supporting the old `overrideAttrs` functionality, while still providing a way to build custom lang tesseract drvs without full recompilation is possible, but very obscure and hacky:
It would involve storing the attrs of `tesseractWithoutData` in its build output via `__structuredAttrs`. The `tesseractWithData` builder could then compare these attrs to its own `__structuredAttrs` and start a compilation (instead of a copy) when attrs differ. If you're interested, I could expand on that.

Currently, I can see only one simple solution to deploy our new features:
Add a new arg (like `withLanguages`) to the tesseract package function.

This stuff will be easy when nixpkgs uses nix-native modules and all `override*` mechanisms are unified, till then it's all rather unsatisfying.

7c6f434c commented:
Well, full compatibility might be hard to reach (but we normally allow this is the refactoring improves other things).

I guess the cleanest appoach would be to have an internal package set with `newScope`, so that one could do things like `tesseract.override { tesseractWithoutData = tesseract.tesseractWithoutData.override {…}; }`

erikarvstedt commented:
Is there an existing package that's defined in a similar way?

7c6f434c commented:
Well, I guess LibreOffice is similar.

erikarvstedt commented:
Sure? `<nixpkgs>/pkgs/applications/office/libreoffice` doesn't contain the word `scope`.

7c6f434c commented:
It's because it is called from the top-level; `newScope` per se is used, say, in `winePackages`.

Or maybe it should just be called a top-level `tesseract-unwrapped` — see, for example, `weechat`

erikarvstedt commented:
If we use the `weechat` approach we're back to `tesseractWithoutData`, now called `tesseract-unwrapped`.

What about just adding a `tesseractWithoutData` arg to the tesseract package function, and also exposing `tesseractWithoutData` via `passthru`?
That would make your above example
```
tesseract.override { tesseractWithoutData = tesseract.tesseractWithoutData.override {…}; }
```
work with minimal effort.

k2pdfopt is the only place in nixpkgs where `tesseract.overrideAttrs` [is used](https://github.com/NixOS/nixpkgs/blob/ace1f064e865fc9e8946da7635ad3229c866ea2c/pkgs/applications/misc/k2pdfopt/default.nix#L78). It could adapted to a similar style:
```
tesseract_modded = tesseract.override {
  tesseractWithoutData = tesseract.tesseractWithoutData.overrideAttrs (attrs: {…})
}
```

7c6f434c commented:
>If we use the `weechat` approach we're back to `tesseractWithoutData`, now called `tesseract-unwrapped`.

Well, yes.

>What about just adding a `tesseractWithoutData` arg to the tesseract package function, and also exposing `tesseractWithoutData` via `passthru`?

I am separately not sure if using unwrapped (or `tesseract.tesseract`) is better for the public interface, given what other packages do.

>k2pdfopt is the only place in nixpkgs where `tesseract.overrideAttrs` is used. It could adapted to a similar style:

Indeed.

erikarvstedt commented:
>  not sure if using unwrapped (or `tesseract.tesseract`) is better for the public interface

In the latest update I've settled for `tesseract.tesseractBase`.

I'm reasonably happy with what we have now. What do you think?

7c6f434c commented on pkgs/applications/graphics/tesseract/default.nix:
```diff
@@ -1,67 +1,83 @@
 
-  enableParallelBuilding = true;
+    # tessdata can be a list of files or a directory containing files
+    inherit tessdata;
```
How does this work without adding `tesseractBase` and `tessdata` to `passthru`?

erikarvstedt replied:
`tesseractBase` is [passed thru](https://github.com/NixOS/nixpkgs/blob/c67bb4ac9c18602ae19913136cae9c8824891dd0/pkgs/applications/graphics/tesseract/default.nix#L79), but via `drv // {…}`.

You're right, `tessdata` modified via `overrideAttrs` gets lost after another `override`. In an earlier draft, `tessdata` was a main package function arg, but I deemed this style too unconvential. I've now restored that patch as it minimizes user errors. 

`tessdata` and `tesseractBase` are now still [attrs of](https://github.com/NixOS/nixpkgs/blob/c36837a80070002f1c286369be7fe794bd2904d0/pkgs/applications/graphics/tesseract/default.nix#L46) `tesseractWithData`, but purely for the nix to builder env variable conversion.

7c6f434c commented on pkgs/applications/graphics/tesseract/default.nix:
```diff
@@ -73,7 +70,7 @@ let
   tesseract = (if enableLanguages == [] then tesseractBase else tesseractWithData) // {
-    inherit languages tesseractBase;
+    inherit tesseractBase tesseractLanguages tessdata;
   };
```
Why not just set `passthru` in `overrideAttrs`?

erikarvstedt replied:
Because these attrs should also be passed thru `tesseractBase`.
We could also define two identical `passthru`s in `tesseractBase` and `tesseractWithData`.
Although redundant, this might improve discoverability for users. What do you prefer?

erikarvstedt replied:
We could also [do this](https://github.com/erikarvstedt/nixpkgs/commit/d5422f0f8ee44d4bfa9bb6eed6e79b40316d213c).

erikarvstedt commented on pkgs/applications/graphics/tesseract/default.nix:
```diff
@@ -1,67 +1,80 @@
+    '';
+  });
+
+  tesseract = (if enableLanguages == [] then tesseractBase else tesseractWithData) // passthru;
```
You mean that `pkgs.tesseract.tesseractBase.tesseractLanguages` doesn't work?
That could be fixed by 
```
passthru = { inherit (tesseractBase // passthru) tesseractLanguages tessdata; };
```
If you meant something else, could you give a short example?

7c6f434c replied:
I am not sure what is the value of `k2pdf.tesseract_modded.tesseractBase`, but I guess it is not supposed to be accessible from outside anyway.

erikarvstedt replied:
Ah, you mean `tesseract.overrideAttrs (_: { tesseractBase = … })`.
Sure, that won't work. That's what I meant by "`tessdata` and `tesseractBase` are attrs of `tesseractWithData` purely for the nix to builder env variable conversion".

erikarvstedt replied:
I could remove the `tessdata` and `tesseractBase` attrs from `tesseractWithData` at the expense 
of more verbose code. Should I do that?

7c6f434c replied:
I think it is not worth it; reasonable cases work, and `overrideAttrs` use to get a private version is not exposed anyway.

erikarvstedt replied:
> I am not sure what is the value of k2pdf.tesseract_modded.tesseractBase
I will work exactly as expected and return the modified base.

7c6f434c replied:
Ah right, there a normal override is donw

7c6f434c commented:
@GrahamcOfBorg build k2pdfopt nixosTests.chromium

erikarvstedt commented:
Very nice. May I squash the fixup commits now?

7c6f434c commented:
Yes, I think so.

erikarvstedt commented:
v4 was still missing, added it now.
I'll do the squashing at the very end.

erikarvstedt commented:
To reassure you, I've changed nothing in the old code, just refactored it to include v4.

7c6f434c commented on pkgs/applications/graphics/tesseract/tesseract4.nix:
```diff
@@ -0,0 +1,27 @@
+    description = "OCR engine";
+    homepage = https://github.com/tesseract-ocr/tesseract;
+    license = stdenv.lib.licenses.asl20;
+    maintainers = with stdenv.lib.maintainers; [viric];
```
I guess it makes sense to add yourself as a maintainer here, too.

7c6f434c commented on pkgs/applications/graphics/tesseract/tesseract4.nix:
```diff
@@ -0,0 +1,27 @@
+    homepage = https://github.com/tesseract-ocr/tesseract;
+    license = stdenv.lib.licenses.asl20;
+    maintainers = with stdenv.lib.maintainers; [viric];
+    platforms = with stdenv.lib.platforms; linux;
```
If tesseract builds don't timeout on ofborg, maybe try darwin here, too?

7c6f434c commented on pkgs/top-level/all-packages.nix:
```diff
@@ -19485,8 +19485,9 @@ in
 
-  tesseract = callPackage ../applications/graphics/tesseract { };
-  tesseract_4 = lowPrio (callPackage ../applications/graphics/tesseract/4.x.nix { });
+  inherit (import ../applications/graphics/tesseract pkgs)
```
I would prefer to pass `callPackage` from the top scope, not `pkgs`.

erikarvstedt replied:
Like so? https://github.com/erikarvstedt/nixpkgs/commit/d7f298d4526bafd50149be668ffd89b0e21e32c7

7c6f434c replied:
Yes

7c6f434c commented:
I think the general style across nixpkgs is to call the wrapper just a wrapper, not put its purpose into names/filenames.

7c6f434c commented:
@GrahamcOfBorg build tesseract_4

7c6f434c commented:
Hm, apparently there is an evaluation error

erikarvstedt commented:
@GrahamcOfBorg build tesseract_4 nixosTests.chromium python3Packages.pyocr

erikarvstedt commented:
pyocr was previously broken on aarch64, this is not related to this PR.

7c6f434c commented on pkgs/applications/graphics/tesseract/get-language-hashes.sh:
```diff
@@ -0,0 +1,44 @@
+  langCodes=$(nix eval --raw '(
+    with (import ../../../.. { config = {}; overlays = []; });
+    let
+      languages = (pkgs.callPackage ./languages.nix {}).v3;
```
Should we allow overriding version for the language list?

erikarvstedt replied:
Thought about that, too, but decided to not make this any more complex.

`--local-langs` will never be used by maintainers. We always want to fetch all languages when updating the tessdata version, 

`--local-langs` is more of a hint for users who want to build a custom language set, not a fully fleshed-out feature.

7c6f434c replied:
Then maybe allow passing a space-separated list of languages as a single argument? All that current-Nixpkgs-list extraction magic seems to lack a use-case…

erikarvstedt replied:
You're right, let's keep this simple.

7c6f434c commented:
A trusted by ofborg user is needed to check Darwin build.
@GrahamcOfBorg build tesseract_4
@GrahamcOfBorg eval

erikarvstedt committed:
tesseract: change file layout

Rename default.nix -> tesseract3.nix
Rename 4.x.nix -> tesseract4.nix

This is needed for the following commits.

erikarvstedt committed:
tesseract: add a wrapper to setup languages

Tesseract is now decoupled from the tessdata language corpus.

This avoids recompilation when building Tesseract with a custom set
of languages.

Update k2pdfopt to use the new wrapper interface.

erikarvstedt committed:
tesseract: add separate language derivations

This frees users from downloading all languages when building
Tesseract with a custom set of languages.

`enableLanguagesHash` is now obsolete.

erikarvstedt commented:
Squashed 'em all.

7c6f434c commented:
Sorry for me only thinking of it now… Does all that binary patching do the same thing as a `wrapProgram` setting `TESSDATA_PREFIX`?

erikarvstedt commented:
The `tesseract` executable could just be wrapped, yes.
But the lib needs binary patching.
Should I wrap the executable?

erikarvstedt commented:
[Commit: Wrap executable](https://github.com/erikarvstedt/nixpkgs/commit/ff789a746802c4bb4c7ad873cc0c8b81ff636a2b)
More complex builder, but we save some space. I'll let you decide.

7c6f434c commented:
Hm. Maybe do multiple outputs? Wrapped executable in `out`, patched library in `lib`, symlinked include files in `dev`.

erikarvstedt commented:
There's also https://github.com/NixOS/nixpkgs/pull/43973, which proposes tesseract `outputs = [ "bin" "lib" "data" "dev" "man" "out" ];`.
Would you prefer that?

erikarvstedt commented:
1. If we split `bin` and `man`, we can't run `man tesseract` inside `nix-shell -p tesseract` (or even from profiles built with `nix-env`, I haven't tested that). So I'd strongly vote against that split.

2. If we split outputs `out` and `lib`, we also have to create an additional `data` output. 
   Reason:  Outputs `out` and `lib` create a cyclic output depency which isn't allowed by the builder:

   `bin/tesseract` in `out` references `lib` in `lib`
   and
   `lib/libtesseract.so.3.0.5` in `lib` references `share` in `out`

   If we have a `data` output, we'd be force to create it (as empty dirs) in the tesseractBase builders.

   We could also patch the autoconf phase to avoid the extra `data` output.
   In any case, we add extra complexity.

3. Even if we have successfully split the `tesseractBase` outputs, the `tesseractWithData` gets still more
   complicated. We have to:
      - copy the `lib` output and patch it.
      - symlink `$dev/lib/include
      - copy and patch `$dev/lib/pkgconfig`, `$dev/nix-support/propagated-build-inputs`.
   
It's all doable, but I don't think we gain much from it.
So I'd propose to keep things as they are.

7c6f434c commented:
I agree, it's more annoying than I expected.

7c6f434c commented:
Are you going to squash the last three commits into the first three?

erikarvstedt commented:
Sorry, I had to reword that comment :cat: 
... and it's squashed again.

7c6f434c commented:
@GrahamcOfBorg build tesseract tesseract4 nixosTexts.chromium python3Packages.pyocr k2pdfopt

erikarvstedt commented:
There were 2 misspellings in the borg command:
`nixosTe(x)ts.chromium`
`tesseract(_)4`

`tesseract_4` is an annoying misnomer. Is it possible to add a `tesseract4` alias?

7c6f434c commented:
@GrahamcOfBorg build tesseract tesseract_4 nixosTests.chromium python3Packages.pyocr k2pdfopt

erikarvstedt committed:
tesseract: rename to tesseract4, add alias

This is more consistent with the naming of the most popular versioned pkgs.

7c6f434c commented:
Well, in normal use all our version naming schemes do not seem to be too annoying… And of course we have all the possible approaches in use.

7c6f434c commented on pkgs/top-level/all-packages.nix:
```diff
@@ -19485,8 +19485,9 @@ in
-  tesseract_4 = lowPrio (callPackage ../applications/graphics/tesseract/4.x.nix { });
+  inherit (callPackage ../applications/graphics/tesseract {})
+    tesseract
+    tesseract_4;
```
Missed that?

erikarvstedt replied:
weirdly, my grep missed that.

erikarvstedt commented:
So better remove that alias again? I'd prefer `tesseract4` as it's more in line with the most popular versioned pkg names.

7c6f434c commented:
Well, alias is OK.

erikarvstedt commented:
Alright, let's do a final borg build.

7c6f434c commented:
Hm. Do you think haveing `tesseract3` and `tesseract` just pointing to `tesseract3` (see: `love`) is a bad idea?

erikarvstedt commented:
Like so? https://github.com/erikarvstedt/nixpkgs/commit/4770ab76dbc0620d3bd1b052c745392ad3858e88
Yeah, that would be fine.

erikarvstedt commented:
Ah, you mean the game engine :smile: 
```
inherit (callPackage ../applications/graphics/tesseract {})
    tesseract3
    tesseract4;
tesseract = tesseract3;
```
That's fine, too, I'll add it.

erikarvstedt committed:
tesseract: add tesseract3 top-level attr

erikarvstedt commented:
Ah, nice screwup. Tesseract 4 expects a different `TESSDATA_PREFIX` format than version 3, that's why the chromium test fails.
I'll be back in an hour, then I'll fix it.

erikarvstedt commented:
@GrahamcOfBorg build tesseract_4 nixosTests.chromium python3Packages.pyocr k2pdfopt

7c6f434c commented:
Is pyocr failure related or not?

erikarvstedt commented:
It's unrelated to this PR.

erikarvstedt commented:
Here's a [similar failure](https://github.com/NixOS/nixpkgs/runs/38699292) from a PR that doesn't affect Tesseract. Unfortunately, I can't read the full logs right now.

7c6f434c commented:
Ah indeed, it is about `cuneiform.py`

erikarvstedt commented:
@GrahamcOfBorg' bring us two beers!

That was quite a ride. Thanks, Michael, for your patience and your great feedback!

7c6f434c commented:
As of patience, it is definitely me who should thank you!

Profpatsch commented:
> 1. If we split `bin` and `man`, we can't run `man tesseract` inside `nix-shell -p tesseract` (or even from profiles built with `nix-env`, I haven't tested that). So I'd strongly vote against that split.

Nope, that’s not true.

```
> du -Lhs /nix/store/kj687mc1zi37yg4c718d5lwla5cs4s47-tesseract-3.05.00
1.1G	/nix/store/kj687mc1zi37yg4c718d5lwla5cs4s47-tesseract-3.05.00
```

This means I have to download >300MB (and extract them to >1GB) just to read the manpages? That seems excessive to me, and was the reason I submitted https://github.com/NixOS/nixpkgs/pull/43973 in the first place.

7c6f434c commented:
>This means I have to download >300MB (and extract them to >1GB) just to read the manpages?

Well, just to read manpages you would download tesseract.tesseractBase (4.5 MiB unpacked)…
